onnxruntime/core/providers/partitioning_utils.cc

```diff
@@ -3,97 +3,173 @@
-#include "core/graph/model.h"
+#include "core/graph/graph_viewer.h"
+#include "core/providers/common.h"
 
 namespace onnxruntime {
 namespace utils {
 
 // internal helpers
 namespace {
-// Kahn's topological sort with awareness of whether a node should be in a supported or unsupported partition
-std::vector<const Node*> PartitionAwareTopoSort(const GraphViewer& graph_viewer,
-                                                const std::unordered_set<const Node*>& supported_nodes) {
-  std::queue<const Node*> supported_to_visit, unsupported_to_visit;
-  std::unordered_map<NodeIndex, size_t> in_degree;
-  std::vector<const Node*> topo_order;
-
-  auto num_nodes = graph_viewer.NumberOfNodes();
-  topo_order.reserve(num_nodes);
-  in_degree.reserve(num_nodes);
-
-  auto add_to_visit = [&](const Node& node) {
-    if (supported_nodes.count(&node)) {
-      supported_to_visit.push(&node);
-    } else {
-      unsupported_to_visit.push(&node);
+#ifndef NDEBUG
+std::string NodeDebugString(const Node& node) {
+  std::ostringstream oss;
+  oss << node.Index() << " '" << node.Name() << "'(" << node.OpType() << ")";
+  return oss.str();
+}
+
+template <typename Container>
+std::string NodeGroupDebugString(const Container& group, bool show_all = false) {
+  static_assert(std::is_same_v<typename Container::value_type, const Node*>);
+
+  if (group.empty()) {
+    return "<no nodes>";
+  }
+
+  std::ostringstream oss;
+  oss << "<" << group.size() << (group.size() == 1 ? " node> " : " nodes> ");
+  if (show_all) {
+    auto node_it = group.begin();
+    oss << NodeDebugString(**(node_it++));
+    while (node_it != group.end()) {
+      oss << ", " << NodeDebugString(**(node_it++));
     }
-  };
+  } else {
+    const Node& start_node = *group.front();
+    const Node& end_node = *group.back();
+    oss << NodeDebugString(start_node) << " to " << NodeDebugString(end_node);
+  }
 
-  // find root nodes
-  for (auto& node : graph_viewer.Nodes()) {
-    size_t input_edge_count = node.GetInputEdgesCount();
-    in_degree.insert({node.Index(), input_edge_count});
-    if (input_edge_count == 0) {
-      add_to_visit(node);
+  return oss.str();
+}
+#endif
+
+std::vector<std::vector<const Node*>> CreateSupportedPartitionNodeGroups(
+    const GraphViewer& graph_viewer,
+    const IsNodeSupportedFn& is_node_supported_fn,
+    const OnGroupClosedFn& on_group_closed_fn,
+    bool debug_output) {
+#ifdef NDEBUG
+  ORT_UNUSED_PARAMETER(debug_output);
+#endif
+
+  ORT_ENFORCE(is_node_supported_fn, "Node support test is required.");
+
+  std::vector<std::vector<const Node*>> supported_groups{};
+
+  // number of inputs from unprocessed nodes (in-degree) per node
+  std::unordered_map<NodeIndex, size_t> in_degree{};
+  // nodes that are ready to process
+  std::deque<const Node*> nodes_to_process{};
+  // nodes that will be processed when considering the next partition node group
+  std::deque<const Node*> nodes_to_process_with_next_group{};
+
```

Comment:
Not clear to me what the 'next partition node group' means. We create a partition from supported nodes and essentially ignore all the others, so conceptually I'm not sure there's an 'unsupported partition'.

Reply:
yes, we only put supported nodes in partitions. in some cases, we will have more than one partition. we form the partitions one at a time so the next partition is just the subsequent one, if any.